node/cn/peer_set_test.go

```diff
@@ -0,0 +1,654 @@
+	cnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.CONSENSUSNODE)).Times(1)
+	cnPeer.EXPECT().GetAddr().Return(addr1).Times(3)
+	cnPeer.EXPECT().GetID().Return(nodeID1.String()).Times(2)
+	cnPeer.EXPECT().Broadcast().AnyTimes()
+	cnPeer.EXPECT().Head().Return(common.Hash{}, big.NewInt(111)).Times(2)
+
+	pnPeer := NewMockPeer(mockCtrl)
+	pnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.PROXYNODE)).Times(1)
+	pnPeer.EXPECT().GetAddr().Return(addr2).Times(3)
+	pnPeer.EXPECT().GetID().Return(nodeID2.String()).Times(2)
+	pnPeer.EXPECT().Broadcast().AnyTimes()
+	pnPeer.EXPECT().Head().Return(common.Hash{}, big.NewInt(222)).Times(2)
+
+	enPeer := NewMockPeer(mockCtrl)
+	enPeer.EXPECT().ConnType().Return(p2p.ConnType(node.ENDPOINTNODE)).Times(2)
+	enPeer.EXPECT().GetAddr().Return(addr3).Times(4)
+	enPeer.EXPECT().GetID().Return(nodeID3.String()).Times(2)
+	enPeer.EXPECT().Broadcast().AnyTimes()
+	enPeer.EXPECT().Head().Return(common.Hash{}, big.NewInt(333)).Times(1)
+	enPeer.EXPECT().Close().Times(1)
+
+	assert.Nil(t, peerSet.BestPeer())
+
+	assert.NoError(t, peerSet.Register(cnPeer))
+	assert.NoError(t, peerSet.Register(pnPeer))
+	assert.NoError(t, peerSet.Register(enPeer))
+
+	assert.Equal(t, enPeer, peerSet.BestPeer())
+
+	assert.NoError(t, peerSet.Unregister(nodeID3.String()))
+
+	assert.Equal(t, pnPeer, peerSet.BestPeer())
+}
+
+func TestPeerSet_Close(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	cnPeer := NewMockPeer(mockCtrl)
+	cnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.CONSENSUSNODE)).Times(1)
+	cnPeer.EXPECT().GetAddr().Return(addr1).Times(3)
+	cnPeer.EXPECT().GetID().Return(nodeID1.String()).Times(2)
+	cnPeer.EXPECT().Broadcast().AnyTimes()
+	cnPeer.EXPECT().DisconnectP2PPeer(p2p.DiscQuitting).Times(1)
+
+	pnPeer := NewMockPeer(mockCtrl)
+	pnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.PROXYNODE)).Times(1)
+	pnPeer.EXPECT().GetAddr().Return(addr2).Times(3)
+	pnPeer.EXPECT().GetID().Return(nodeID2.String()).Times(2)
+	pnPeer.EXPECT().Broadcast().AnyTimes()
+	pnPeer.EXPECT().DisconnectP2PPeer(p2p.DiscQuitting).Times(1)
+
+	enPeer := NewMockPeer(mockCtrl)
+	enPeer.EXPECT().ConnType().Return(p2p.ConnType(node.ENDPOINTNODE)).Times(1)
+	enPeer.EXPECT().GetAddr().Return(addr3).Times(3)
+	enPeer.EXPECT().GetID().Return(nodeID3.String()).Times(2)
+	enPeer.EXPECT().Broadcast().AnyTimes()
+	enPeer.EXPECT().DisconnectP2PPeer(p2p.DiscQuitting).Times(1)
+
+	assert.NoError(t, peerSet.Register(cnPeer))
+	assert.NoError(t, peerSet.Register(pnPeer))
+	assert.NoError(t, peerSet.Register(enPeer))
+
+	assert.False(t, peerSet.closed)
+	peerSet.Close()
+	assert.True(t, peerSet.closed)
+}
+
+func TestPeerSet_SampleResendPeersByType_PN(t *testing.T) {
+	// CN Peer=1, PN Peer=0
+	{
+		peerSet := newPeerSet()
+		mockCtrl := gomock.NewController(t)
+
+		cnPeer := NewMockPeer(mockCtrl)
+		cnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.CONSENSUSNODE)).Times(2)
```

Comment:
The code to initialize the mocke peer seems to be repeated a lot. How about creating a function?

Reply:
Good point =) As you suggested, I will make a function which receives parameters of expected times.